pycernan/avro/client.py

```diff
@@ -1,47 +1,126 @@
 """
     Base Avro client from which all other clients derive.
 """
+import contextlib
 import socket
 
 import pycernan.avro.config
 
 from abc import ABCMeta, abstractmethod
+from queue import Queue, Empty
 
-from pycernan.avro.exceptions import EmptyBatchException
+from pycernan.avro.exceptions import EmptyBatchException, EmptyPoolException
 from pycernan.avro.serde import serialize
 
 
+_DefunctConnection = object()
+
+
+class TCPConnectionPool(object):
+    """
+    Fork friendly TCP connection pool.
+
+    Adapted from: https://github.com/gevent/gevent/blob/master/examples/psycopg2_pool.py
+    """
+    def __init__(self, host, port, maxsize, connect_timeout, read_timeout):
+        if maxsize <= 0:
+            raise ValueError("maxsize must be > 0")
+
+        self.host = host
+        self.port = port
+        self.size = 0
+        self.maxsize = maxsize
+
+        self.connect_timeout = connect_timeout
+        self.read_timeout = read_timeout
+        self.pool = Queue()
+        for _ in range(maxsize):
+            self._put(_DefunctConnection)
+
+    def _create_connection(self):
+        sock = socket.create_connection((self.host, self.port), timeout=self.connect_timeout)
+        sock.settimeout(self.read_timeout)
+        return sock
+
+    def _get(self, _block=True):
+        try:
+            connection = self.pool.get(_block)
+        except Empty:
+            # Expected to happen only when users override _block=False
+            raise EmptyPoolException()
+
+        # When a connection is defunct, we attempt to
+        # regenerate it.  Note - it is important that we always return
+        # something back to the queue here so that we don't errode the
+        # errode our capacity.
+        if connection is _DefunctConnection:
+            try:
+                connection = self._create_connection()
+            except:
+                # Always return a resource to the queue,
+                # even if it is defunct.
+                self._put(_DefunctConnection)
+                raise
+
+        return connection
+
+    def _put(self, item):
+        # Why this function?  It makes mocking in unit tests easier.
+        self.pool.put(item)
+
+    def closeall(self):
+        """
+            Close established connections currently not in-use.
+        """
+        try:
+            while True:
+                conn = self.pool.get_nowait()
+                if conn is _DefunctConnection:
+                    continue
+                conn.close()
+        except Exception:
+            pass
+
+    @contextlib.contextmanager
+    def connection(self, _block=True):
+        """
+            Context manager for pooled connections.
+        """
+        conn = self._get(_block)
+        try:
+            yield conn
+        except:
+            conn = _DefunctConnection
```

Comment:
So the idea here is to discard bad connections so they get reestablished the next time someone tries to get a connection right? As written, this happens for _any_ kind of exception, even ones unrelated to the connection.

On one hand, maybe better safe than sorry?

On the other hand, I could see this leading to cascading failures where an application logic error causes a large number of connections to be considered "defunct". Then the high rate of DNS lookups in establishing new connections overloads kube-dns, or the overhead of establishing new TCP connections overloads the server, etc.

I also note if the exception was not due to a connection failure, the connection is not gracefully closed here. It will probably get closed quickly when the refcount drops to zero, but if it's involved in any reference cycle we might accumulate unused connections until the GC cleans them up, perhaps accumulating enough old connections before then to run out of file descriptors, etc.

Would it be more prudent to be more selective about the kinds of exceptions that will taint the connection, or should there be some kind of health checking of the connection that occurs before it is reused?

Reply:
In this case, the application layer may be in the middle of a transaction (of sorts) and raise an exception. 

Motivating example - a pycernan publisher publishes some data, waits for an ack, and times-out.  That data may arrive later if the connection isn't closed and result in issues for a subsequent publisher when they receive an ack for a publish they didn't send.  Pycernan currently raises an Exception when acks arrive out of order.

As for connection thrashing - 

I feel like for a library like this, leaving etiquette up to the user wrt connect thrashing is common practice. We "do the right thing" by my estimation in that we raise all exceptions back to the user to deal with.  If operational issues arise, they can respond how they want without this library being too opinionated or adding complicated retry policy constructs.

As for closing the connection -  I agree.  I will add logic here to close the connection from the client side.